ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/resources/arjunacore/XAResourceRecord.java

```diff
@@ -212,13 +212,11 @@ public int topLevelPrepare()
 			 * XAER_PROTO.
 			 */
 
-			if (_rollbackOptimization) // won't have rollback called on it
-				removeConnection();
```

Comment:
@jmfinelli when_rolledBack = true, from topLevelPrepare() are you sure that removeConnection() (perhaps it will be https://github.com/jbosstm/narayana/pull/2214/commits/5f0fdfeb59f119f604ceb215baa2a72609294eb8#diff-6ba9616ad040b9a0f852604d2d443821d786bb2982a9d6a1ad0bd6df048f8730R773) is called? If it's not called a transactional driver `close()` opportunity is lost so I would like to know.

Reply:
I think so: `removeConnection()` should cover all cases.

Reply:
Thank you!

Reply:
Please can you clarify why it was safe to remove a conditional call to removeConnection?

Reply:
So I expect it is close to something like _rollbackOptimization was default to false, and given it's no longer configurable it's expected to not be needed. Personally I don't think that deprecating something, implies removing the functionality.

Reply:
Also, in your comment you said, removeConnection should cover all the case where the record hasn't been prepared. Please can you point out the code where removeConnection is called when the prepare has not completed for the record?

Reply:
I do understand though, we can't call removeConnection if we get XAER_INVAL, XAER_PROTO, XAER_RMERR or XAER_RMFAIL. So maybe this removeConnection just moves to after the line below where `_rolledBack = true;`?

Reply:
BTW when running the current changes and the test removeConnection did not get called (for either TestResource)

Reply:
Looking at the code, I am not seeing how removeConnection is called for prepared resources in the rollback case (already today) as it is guarded by !_prepared (unlike the commit case which does not check for whether it was _prepared or not)

Reply:
Transactional driver seems to be able to release connections even when removeConnection is not called. I will resolve this